src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,315 @@
+	// to all incoming regular expressions to ensure compatibility between them.
+
+	// first, we parse the regular expression into the equivalent regex
+	reString := string(r)
+	reAst, err := parseRegexp(reString)
+	if err != nil {
+		return CompiledRegex{}, err
+	}
+
+	// Issue (a): Vellum does not allow regexps which use characters '^', or '$'.
+	// To address this issue, we strip these characters from appropriate locations in the parsed syntax.Regexp
+	// for Vellum's RE.
+	vellumRe, err := ensureRegexpUnanchored(reAst)
+	if err != nil {
+		return CompiledRegex{}, fmt.Errorf("unable to create FST re: %v", err)
+	}
+
+	// Issue (b): Vellum treats every regular expression as anchored, where as the map-backed segment does not.
+	// To address this issue, we ensure that every incoming regular expression is modified to be anchored
+	// when querying the map-backed segment, and isn't anchored when querying Vellum's RE.
+	simpleRe, err := ensureRegexpAnchored(vellumRe)
+	if err != nil {
+		return CompiledRegex{}, fmt.Errorf("unable to create map re: %v", err)
+	}
+
+	simpleRE, err := re.Compile(simpleRe.String())
+	if err != nil {
+		return CompiledRegex{}, err
+	}
+	compiledRegex := CompiledRegex{Simple: simpleRE}
+
+	fstRE, start, end, err := fstregexp.ParsedRegexp(reString, vellumRe)
+	if err != nil {
+		return CompiledRegex{}, err
+	}
+	compiledRegex.FST = fstRE
+	compiledRegex.PrefixBegin = start
+	compiledRegex.PrefixEnd = end
+
+	return compiledRegex, nil
+}
+
+func parseRegexp(re string) (*syntax.Regexp, error) {
+	return syntax.Parse(re, syntax.Perl)
+}
+
+// ensureRegexpAnchored adds '^' and '$' characters to appropriate locations in the parsed syntax.Regexp,
+// to ensure every input regular expression is converted to it's equivalent anchored regular expression.
+// NB: assumes input regexp AST is un-anchored.
+func ensureRegexpAnchored(unanchoredRegexp *syntax.Regexp) (*syntax.Regexp, error) {
+	ast := &syntax.Regexp{
+		Op:    syntax.OpConcat,
+		Flags: syntax.Perl,
+		Sub: []*syntax.Regexp{
+			&syntax.Regexp{
+				Op:    syntax.OpBeginText,
+				Flags: syntax.Perl,
+			},
+			unanchoredRegexp,
+			&syntax.Regexp{
+				Op:    syntax.OpEndText,
+				Flags: syntax.Perl,
+			},
+		},
+	}
+	return simplify(ast), nil
+}
+
+// ensureRegexpUnanchored strips '^' and '$' characters from appropriate locations in the parsed syntax.Regexp,
+// to ensure every input regular expression is converted to it's equivalent un-anchored regular expression
+// assuming the entire input is matched.
+func ensureRegexpUnanchored(parsed *syntax.Regexp) (*syntax.Regexp, error) {
+	r, _, err := ensureRegexpUnanchoredHelper(parsed, true, true)
+	if err != nil {
+		return nil, err
+	}
+	return simplify(r), nil
+}
+
+func ensureRegexpUnanchoredHelper(parsed *syntax.Regexp, leftmost, rightmost bool) (output *syntax.Regexp, changed bool, err error) {
+	newRegexp := parsed
+	changed = false
+	switch parsed.Op {
```

Comment:
Presumably this switch statement needs to handle every type of A.S.T node that we could encounter. Is it worth adding a default that returns an error so that regex's we didn't anticipate properly at least 500 or something instead of having incorrect behavior